src/platforms/eglstream-kms/server/egl_output.cpp

```diff
@@ -316,6 +275,67 @@ uint32_t mgek::EGLOutput::crtc_id() const
     return static_cast<uint32_t>(crtc_id);
 }
 
+int mgek::EGLOutput::atomic_commit(uint64_t fb, const void *drm_event_userdata, uint32_t flags)
+{
+    auto const width = connector->modes[mode_index].hdisplay;
+    auto const height = connector->modes[mode_index].vdisplay;
+
+    std::unique_ptr<drmModeAtomicReq, void(*)(drmModeAtomicReqPtr)>
+        request{drmModeAtomicAlloc(), &drmModeAtomicFree};
+
+    if (flags & DRM_MODE_ATOMIC_ALLOW_MODESET)
+    {
+        mgk::ObjectProperties crtc_props{drm_fd, _crtc_id, DRM_MODE_OBJECT_CRTC};
+
+        /* Activate the CRTC and set the mode */
+        drmModeAtomicAddProperty(request.get(), _crtc_id, crtc_props.id_for("MODE_ID"), mode_id);
+        drmModeAtomicAddProperty(request.get(), _crtc_id, crtc_props.id_for("ACTIVE"), 1);
+
+        /* Set CRTC for the output */
+        auto const connector_id = connector->connector_id;
+        mgk::ObjectProperties connector_props{drm_fd, connector_id, DRM_MODE_OBJECT_CONNECTOR};
+        drmModeAtomicAddProperty(request.get(), connector_id, connector_props.id_for("CRTC_ID"), _crtc_id);
+    }
+
+    /* Source viewport. Coordinates are 16.16 fixed point format */
+    drmModeAtomicAddProperty(request.get(), plane_id, plane_props->id_for("SRC_X"), 0);
+    drmModeAtomicAddProperty(request.get(), plane_id, plane_props->id_for("SRC_Y"), 0);
+    drmModeAtomicAddProperty(request.get(), plane_id, plane_props->id_for("SRC_W"), width << 16);
+    drmModeAtomicAddProperty(request.get(), plane_id, plane_props->id_for("SRC_H"), height << 16);
+
+    /* Destination viewport. Coordinates are *not* 16.16 */
+    drmModeAtomicAddProperty(request.get(), plane_id, plane_props->id_for("CRTC_X"), 0);
+    drmModeAtomicAddProperty(request.get(), plane_id, plane_props->id_for("CRTC_Y"), 0);
+    drmModeAtomicAddProperty(request.get(), plane_id, plane_props->id_for("CRTC_W"), width);
+    drmModeAtomicAddProperty(request.get(), plane_id, plane_props->id_for("CRTC_H"), height);
+
+    /* Set a surface for the plane */
+    drmModeAtomicAddProperty(request.get(), plane_id, plane_props->id_for("CRTC_ID"), _crtc_id);
+    drmModeAtomicAddProperty(request.get(), plane_id, plane_props->id_for("FB_ID"), fb);
+
+    /* We don't monitor the DRM events (yet), so have no userdata */
+    auto ret = drmModeAtomicCommit(drm_fd, request.get(), flags, const_cast<void*>(drm_event_userdata));
+
+    if (ret != 0)
+    {
+        return ret;
+    }
+
+    using_saved_crtc = false;
+    return 0;
+}
+
+bool mgek::EGLOutput::queue_atomic_flip(FBHandle const& fb, void const* drm_event_userdata)
+{
+    int ret = atomic_commit(fb, drm_event_userdata, DRM_MODE_ATOMIC_NONBLOCK | DRM_MODE_PAGE_FLIP_EVENT);
+    if (ret != 0)
+    {
+        BOOST_THROW_EXCEPTION((std::system_error{-ret, std::system_category(), "Failed to commit Atomic KMS state"}));
```

Comment:
So, there are a bunch of ways `atomic_commit` can fail, but transiently. It can require a modeset after switching back from VT switching. It can also fail with `EBUSY` if there's already a pending commit that hasn't finished (although we shouldn't hit that, having waited for previous page flips.)

Maybe we don't have to handle `EBUSY` here, and just assume that our page-filp handling will cover it.

We *should* handle the resume-from-VT-switch, which probably means twiddling a flag in `mge::Display::resume()` that causes `queue_atomic_flip` to additionally pass `ALLOW_MODESET`?

Reply:
I implemented something that makes a lot of sense, but I get a fatal error whenever I VT switch :thinking: